src/openvic-simulation/scripts/Condition.hpp

```diff
@@ -0,0 +1,81 @@
+#pragma once
+
+#include "openvic-simulation/types/IdentifierRegistry.hpp"
+#include <variant>
+#include <vector>
+
+namespace OpenVic {
+	struct ConditionManager;
+
+	// Condition: the definition of an individual condition or scope
+	// AND, OR and NOT are conditions with complex type
+	// Scopes are treated as ANDs
+	struct Condition : HasIdentifier {
+		friend struct ConditionManager;
+
+		enum class value_type_t {
+			IDENTIFIER,
+			STRING,
+			BOOLEAN,
+			INTEGER,
+			REAL,
+			COMPLEX
+		};
+
+		enum class scope_t {
+			GLOBAL,
+			COUNTRY,
+			PROVINCE,
+			POP
+		};
+
+	private:
+		const value_type_t PROPERTY(value_type);
+		std::set<scope_t> PROPERTY(scopes);
```

Comment:
You could use a bitfield for this, so:
```c++
enum scope_t : uint8_t {
	COUNTRY  = 1 << 0, // 0b001
	PROVINCE = 1 << 1, // 0b010
	POP      = 1 << 2, // 0b100
	GLOBAL   = COUNTRY | PROVINCE | POP // 0b111
}
...
scope_t PROPERTY(scopes);
```
It would be faster to modify and to check what scopes are enabled, e.g. `(scopes & new_scope) == new_scope` checks `scopes` contains all of `new_scope`, you could also just check the result is non-zero but that wouldn't work for checking if it's `GLOBAL` or a combination of 2. Also it would treat `GLOBAL` as the same as adding all 3 others individually, and make it so you just have to check `scopes | PROVINCE` instead of `scopes.contains(PROVINCE) || scopes.contains(GLOBAL)`. `class` is removed from the enum declaration because otherwise you can't use bitwise operators on it, e.g. `scope |= scope_t::PROVINCE;` (with an `enum class` you'd have to `static_cast` to `uint8_t` every time you want to use `&` or `|`).

Reply:
There's also `std::bitset` which is designed for this kind of stuff, although it wouldn't have the same type as the enums so that might cause complications.